File: 2021/src/day03.rs
use crate::utils;

pub fn day03() {
    let binary_diagnostic = utils::read_input_file("src/day03_input.txt")
        .unwrap()
        .iter()
        .map(|x| x.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    let mut gamma_rate = Vec::new();
    let mut epsilon_rate = Vec::new();

    for x in 0..binary_diagnostic[0].len() {
        let mut zeros = 0;
        let mut ones = 0;
        for y in 0..binary_diagnostic.len() {
            if (binary_diagnostic[y][x] == '0') {
                zeros = zeros + 1;
            } else {
                ones = ones + 1;
            }
        }

        let max_min = if ones > zeros { ["0", "1"] } else { ["1", "0"] };

        gamma_rate.push(max_min[0]);
        epsilon_rate.push(max_min[1])
    }

    let gamma_rate_value = isize::from_str_radix(&gamma_rate.join(""), 2).unwrap();
    let epsilon_rate_value = isize::from_str_radix(&epsilon_rate.join(""), 2).unwrap();

    println!("{}", gamma_rate_value * epsilon_rate_value);
}

pub fn day03_part2() {
    let binary_diagnostic = utils::read_input_file("src/day03_input.txt")
        .unwrap()
        .iter()
        .map(|x| x.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    let mut gamma_rate = Vec::new();
    let mut epsilon_rate = Vec::new();

    for x in 0..binary_diagnostic[0].len() {
        let mut zeros = 0;
        let mut ones = 0;
        for y in 0..binary_diagnostic.len() {
            if (binary_diagnostic[y][x] == '0') {
                zeros = zeros + 1;
            } else {
                ones = ones + 1;
            }
        }

        let max_min = if ones > zeros { ["0", "1"] } else { ["1", "0"] };

        gamma_rate.push(max_min[0]);
        epsilon_rate.push(max_min[1])
    }

    let gamma_rate_value = isize::from_str_radix(&gamma_rate.join(""), 2).unwrap();
    let epsilon_rate_value = isize::from_str_radix(&epsilon_rate.join(""), 2).unwrap();

    println!("{}", gamma_rate_value * epsilon_rate_value);
}

File: 2021/dart/lib/day03.dart
import 'utils.dart' as utils;

int day03() {
  var binaryDiagnostic = utils.readFile("../inputs/day03.txt");
  var gammaRate = "";
  var epsilonRate = "";

  for (int x = 0; x < binaryDiagnostic[0].length; x++) {
    var zeros = 0;
    var ones = 0;
    for (int y = 0; y < binaryDiagnostic.length; y++) {
      if (binaryDiagnostic[y][x] == "0") {
        zeros++;
      } else {
        ones++;
      }
    }

    var maxMin = ones > zeros ? ["0", "1"] : ["1", "0"];
    gammaRate += maxMin[0];
    epsilonRate += maxMin[1];
  }

  return utils.convertBinaryStringToDecimal(gammaRate) *
      utils.convertBinaryStringToDecimal(epsilonRate);
}

int getRating(List<String> list, int charIndex, bool lookMax) {
  List<String> ones = [];
  List<String> zeros = [];

  if (list.length == 1) {
    return int.parse(list[0], radix: 2);
  }

  for (int y = 0; y < list.length; y++) {
    var current = list[y];
    if (list[y][charIndex] == "0") {
      zeros.add(current);
    } else {
      ones.add(current);
    }
  }

  List<String> choosenArray = [];

  if (lookMax) {
    choosenArray = ones.length >= zeros.length ? ones : zeros;
  } else {
    choosenArray = zeros.length <= ones.length ? zeros : ones;
  }
  return getRating(choosenArray, charIndex + 1, lookMax);
}

int day03Part2() {
  var binaryDiagnostic = utils.readFile("../inputs/day03.txt");

  var oxygenGenerator = getRating(binaryDiagnostic, 0, true);
  var co2Scrubber = getRating(binaryDiagnostic, 0, false);

  return oxygenGenerator * co2Scrubber;
}
